### ISSUE TITLE
bumping ecto version from 2.1.6 to 2.2.0 causes key :action not found error

### ISSUE THREAD

**Issue by atomkirk** [elixir-ecto/ecto#2178]
### Environment

* Elixir version (elixir -v): 1.5.1
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): mysql 5.7.18
* Ecto version (mix deps): 2.2.0
* Database adapter and version (mix deps): mariaex 0.8.3
* Operating system: OS X Sierra

### Current behavior

I have the following changeset:
```
  def changeset(struct, params \\ %{}) do
    struct
    |> cast(params, allowed_params())
    |> cast_assoc(:journal_entry_lines, required: true)
  end
```

This throws the error:

```
** (KeyError) key :action not found in: %ZB.JournalEntryLine{__meta__: …
```

If I remove the `required: true` then most of the tests that failed with it pass, but a few still fail. So it definitely affects things although I don't know thats entirely it.

### Expected behavior

If I specify ecto 2.1.6 in my mix file, this works and I don't see this error. If I change it to 2.2.0, I get this error.

I'm sorry I don't know what more to provide to be helpful. Hopefully if you ask questions I can help you debug this. I don't *think* I'm doing anything fancy (aka wrong) so it seems like this is a bug…

**Comment by josevalim** [comment]
Can you please provide the full stacktrace?

**Comment by josevalim** [comment]
Also, please provide the changeset implementation of `ZB.JournalEntryLine`. I suspect it is returning a struct and not a changeset.

**Comment by hl** [comment]
Do you use `import Ecto` somewhere?

**Comment by atomkirk** [comment]
@josevalim sorry I should have thought to provide the full stacktrace

```
** (KeyError) key :action not found in: %ZB.JournalEntryLine{__meta__: #Ecto.Schema.Metadata<:loaded, "journal_entry_lines">, account: #Ecto.Association.NotLoaded<association :account is not loaded>, account_id: 1597, amount: #Decimal<100.00000000000>, archived_at: nil, bill_items: #Ecto.Association.NotLoaded<association :bill_items is not loaded>, bill_to_contact: nil, bill_to_customer_id: nil, chart_account: %ZB.ChartAccount{__meta__: #Ecto.Schema.Metadata<:loaded, "chart_accounts">, account: #Ecto.Association.NotLoaded<association :account is not loaded>, account_id: 1597, archived_at: nil, bank_account: #Ecto.Association.NotLoaded<association :bank_account is not loaded>, bank_account_id: nil, cache_balance: #Decimal<0.00>, classification: "Liabilities", code: "2000", created_at: #DateTime<2017-08-22 00:00:00.000000Z>, deleted_at: nil, id: 106328, is_system: true, journal_entry_lines: #Ecto.Association.NotLoaded<association :journal_entry_lines is not loaded>, name: "Liabilities", parent: #Ecto.Association.NotLoaded<association :parent is not loaded>, parent_id: nil, start_date: ~D[2017-08-22], sub_code: nil, system_code: "2000", updated_at: #DateTime<2017-08-22 22:59:10.000000Z>, viewable: true}, chart_account_id: 106328, created_at: #DateTime<2017-08-22 22:59:10.000000Z>, currency_code: "USD", date: ~D[2017-08-22], deleted_at: nil, id: 3804, imported_transaction: nil, imported_transaction_id: nil, is_1099_required: false, journal_entry: %ZB.JournalEntry{__meta__: #Ecto.Schema.Metadata<:loaded, "journal_entries">, account: #Ecto.Association.NotLoaded<association :account is not loaded>, account_id: 1597, archived_at: nil, cache_total: #Decimal<100.00>, cloud_files: #Ecto.Association.NotLoaded<association :cloud_files is not loaded>, created_at: #DateTime<2017-08-22 22:59:10.000000Z>, currency_code: "USD", date: ~D[2017-10-07], deleted_at: nil, id: 1502, is_closing: false, is_confirmed: false, is_system: true, journal_entry_lines: #Ecto.Association.NotLoaded<association :journal_entry_lines is not loaded>, linked_journal_entry: #Ecto.Association.NotLoaded<association :linked_journal_entry is not loaded>, linked_journal_entry_id: nil, name: "Amet autem mollitia occaecati ut.", note: "Autem dolore similique!", updated_at: #DateTime<2017-08-22 22:59:10.000000Z>}, journal_entry_id: 1502, kind: :debit, name: "Reiciendis molestias facilis nostrum asperiores.", suggested_chart_account: #Ecto.Association.NotLoaded<association :suggested_chart_account is not loaded>, suggested_chart_account_id: nil, tags: [%ZB.Tag{__meta__: #Ecto.Schema.Metadata<:loaded, "tags">, account: #Ecto.Association.NotLoaded<association :account is not loaded>, account_id: 1597, created_at: #DateTime<2017-08-22 22:59:10.000000Z>, deleted_at: nil, id: 4250, is_system: false, journal_entry_line: #Ecto.Association.NotLoaded<association :journal_entry_line is not loaded>, journal_entry_line_id: 3804, taggable_id: 359, taggable_type: "Label", updated_at: #DateTime<2017-08-22 22:59:10.000000Z>}, %ZB.Tag{__meta__: #Ecto.Schema.Metadata<:loaded, "tags">, account: #Ecto.Association.NotLoaded<association :account is not loaded>, account_id: 1597, created_at: #DateTime<2017-08-22 22:59:10.000000Z>, deleted_at: nil, id: 4251, is_system: false, journal_entry_line: #Ecto.Association.NotLoaded<association :journal_entry_line is not loaded>, journal_entry_line_id: 3804, taggable_id: 360, taggable_type: "Label", updated_at: #DateTime<2017-08-22 22:59:10.000000Z>}], updated_at: #DateTime<2017-08-22 22:59:10.000000Z>}
     stacktrace:
       (ecto) lib/ecto/changeset/relation.ex:32: anonymous fn/1 in Ecto.Changeset.Relation.empty?/2
       (elixir) lib/enum.ex:2787: Enum.all_list/2
       (ecto) lib/ecto/changeset.ex:764: Ecto.Changeset.missing_relation/6
       (ecto) lib/ecto/changeset.ex:725: Ecto.Changeset.cast_relation/4
       (zipbooks) lib/zipbooks/journal_entries/journal_entry.ex:56: ZB.JournalEntry.changeset/2
       (zipbooks) lib/zipbooks_web/controllers/journal_entry_controller.ex:110: ZB.JournalEntryController.update/4
       (zipbooks) lib/zipbooks_web/controllers/journal_entry_controller.ex:1: ZB.JournalEntryController.action/2
       (zipbooks) lib/zipbooks_web/controllers/journal_entry_controller.ex:1: ZB.JournalEntryController.phoenix_controller_pipeline/2
       (zipbooks) lib/zipbooks_web/endpoint.ex:1: ZB.Endpoint.instrument/4
       (phoenix) lib/phoenix/router.ex:278: Phoenix.Router.__call__/1
       (zipbooks) lib/plug/error_handler.ex:64: ZB.Router.call/2
       (zipbooks) lib/zipbooks_web/endpoint.ex:1: ZB.Endpoint.plug_builder_call/2
       (zipbooks) lib/plug/debugger.ex:99: ZB.Endpoint."call (overridable 3)"/2
       (zipbooks) lib/zipbooks_web/endpoint.ex:1: ZB.Endpoint.call/2
       (phoenix) lib/phoenix/test/conn_test.ex:224: Phoenix.ConnTest.dispatch/5
       (zipbooks) test/support/also/test_case.ex:75: Also.TestCase.send_req/2
       (zipbooks) test/support/also/utils.ex:55: Also.Utils.make_request/2
       (zipbooks) test/support/also/helpers.ex:41: anonymous fn/3 in Also.Helpers.assert_data_has_one/3
       (zipbooks) test/support/also/helpers.ex:351: anonymous fn/1 in Also.Helpers.in_transaction/2
       (ecto) lib/ecto/adapters/sql.ex:576: anonymous fn/3 in Ecto.Adapters.SQL.do_transaction/3
```

Here is`ZB.JournalEntryLines.changeset`:

```
  def changeset(struct, params \\ %{}) do
    struct
    |> base_changeset(params)
    |> put_bill_items
    |> propagate(:tags, :account_id)
    |> cast_assoc(:tags)
    |> validate_on_account(:tags, ZB.Tag)
    |> validate_on_account(:chart_account_id, ZB.ChartAccount)
    |> validate_on_account(:bill_to_contact, ZB.Contact)
    |> validate_required(required_fields())
    |> validate_amount
    |> validate_tags
    |> validate_inclusion(:kind, [:debit, :credit], message: "Kind must be debit or credit")
  end
```

**Comment by atomkirk** [comment]
@hl yes, but most notably where in the file phoenix generated for me:
![](http://d.pr/i/9jmKkl/3RpOF6eh+)

**Comment by josevalim** [comment]
Fixed in the v2.2 branch. When we do a new release, you should be covered. Thanks for the report!
